### PR TITLE
Feature/decor column sql (#1288)

### PR DESCRIPTION
**Pull request by cybermerlin**
+ #1288 
+ aliasName for column
+ fix connection|manager .connect|close
+ fix test-utils n tests (trouble with float connection brake|lost while .save() or moving between tests)
+ add equals for compare obj|arr with order by id|*
+ prepare for parallelizm (for build|execution test)
---------------
task accomplished.
I started next task in my repo (my plan in `CHANGELOG.md`)

### CODE REVIEW

**Comment by pleerock** [comment]
Thank you very much for your hard efforts spent on this issue implementation. 

First, its a bit complicated to review PR when there are unrelated changes (code clean, fixes in tests, etc.) it would be great if you do separate PRs this or next times.

Second, we should have a discussion about this feature design first. I don't think this functionality should use `@Column` decorator. It will be better I think to implement this way:

```ts
@ComputedColumn(`...SQL...`, { ...options })
```

**Comment by cybermerlin** [comment]
1. ok, in next time I will be careful with commits, but for one task to use many separate PR - the sound is strange n maybe will a little hard for a developer. Of course u always right n I accepted ur pleasures. Do not a problem.

2. @Column - that's very good decision for this task because I've code generator from DB (pg) into mine js code, and this place (@Column) I use for SQL-code only for a column.
In the next task, I'll be adding something more complex and will choosing another place, because in my next task it's more correct (or useful). And, of course, I'll be with great pleasures looking and trying to resolve all wishes from all people (if I'll be having time).

3. about 
```js
createQueryBuilder()
      .selectAndMap("COUNT(*)", "cntProperty")
```
it's good sound but not now, because I've not time for the extended decision my task, sorry about.

4. I'll be complete this test to approve this PR and I've to hope you approve...

File: src/connection/Connection.ts [hunk 1]
```diff
@@ -152,8 +151,7 @@ export class Connection {
      * but it also can setup a connection pool with database to use.
      */
     async connect(): Promise<this> {
-        if (this.isConnected)
-            throw new CannotConnectAlreadyConnectedError(this.name);
+        if (this.isConnected) return this;
```
Comment (pleerock) [inline src/connection/Connection.ts]: hey, you cannot simply take and change exist logic this way. This change is out of scope of what you are doing, please remove it.

File: src/connection/ConnectionManager.ts [hunk 2]
```diff
@@ -54,7 +53,7 @@ export class ConnectionManager {
 
             // if connection is registered and its not closed then throw an error
             if (existConnection.isConnected)
-                throw new AlreadyHasActiveConnectionError(options.name || "default");
+                return existConnection;
```
Comment (pleerock) [inline src/connection/ConnectionManager.ts]: hey, you cannot simply take and change exist logic this way. This change is out of scope of what you are doing, please remove it.

File: src/driver/mysql/MysqlDriver.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+///<reference path="MysqlQueryRunner.ts"/>
```
Comment (pleerock) [inline src/driver/mysql/MysqlDriver.ts]: what is this?

File: src/driver/mysql/MysqlDriver.ts [hunk 4]
```diff
@@ -202,6 +203,8 @@ export class MysqlDriver implements Driver {
             this.pool = await this.createPool(this.createConnectionOptions(this.options, this.options));
             this.database = this.options.database;
         }
+
+        (new MysqlQueryRunner(this)).query("SET SESSION sql_mode = 'PIPES_AS_CONCAT'");
```
Comment (pleerock) [inline src/driver/mysql/MysqlDriver.ts]: what is this? Please remove.

File: src/metadata/ColumnMetadata.ts [hunk 5]
```diff
@@ -286,6 +303,13 @@ export class ColumnMetadata {
         }
         if (options.args.options.transformer)
             this.transformer = options.args.options.transformer;
+        if (options.args.options.sql) {
+            this.sql = `(${(typeof options.args.options.sql === "string")
+                ? options.args.options.sql
+                : options.args.options.sql.call(this)})`; // TODO: вместо   this   нужен scope?       embeddedMetadata
```
Comment (pleerock) [inline src/metadata/ColumnMetadata.ts]: I dont think someone else knows russian.

File: src/metadata/ColumnMetadata.ts [hunk 6]
```diff
@@ -480,20 +503,22 @@ export class ColumnMetadata {
         }
     }
 
-    // ---------------------------------------------------------------------
-    // Builder Methods
-    // ---------------------------------------------------------------------
+    // endregion
```
Comment (pleerock) [inline src/metadata/ColumnMetadata.ts]: what is this? Please return back original formatting.

File: test/utils/test-utils.ts [hunk 7]
```diff
@@ -215,7 +216,7 @@ export async function createTestingConnections(options?: TestingOptions): Promis
  * Closes testing connections if they are connected.
  */
 export function closeTestingConnections(connections: Connection[]) {
-    return Promise.all(connections.map(connection => connection.isConnected ? connection.close() : undefined));
+    return Promise.all((connections || []).map(connection => connection.isConnected ? connection.close() : undefined));
```
Comment (pleerock) [inline test/utils/test-utils.ts]: does not look like connections can be undefined

File: src/metadata/ColumnMetadata.ts [hunk 8]
```diff
@@ -307,8 +331,7 @@ export class ColumnMetadata {
     }
 
     // ---------------------------------------------------------------------
-    // Public Methods
-    // ---------------------------------------------------------------------
+    // region Public Methods
```
[the remark this answers is not in the document]
Comment (pleerock) [reply src/metadata/ColumnMetadata.ts]: changing code style is not of the current task's scope. Revert back.

**Comment by pleerock** [comment]
I don't want to re-implement things when `.selectAndMap` will be introduced, so I prefer to keep this PR opened until it will be. If you don't have a time maybe someone else can continue to work on it. Also as I told its not a good idea to use Column decorator, instead we shall create a new decorator.

**Comment by cybermerlin** [comment]
Column is very comfortable place for solve this task, because is too more readable, n nobody need know about exists something else for define he columns. That's just column n nothing else, no additional Select, no something else, just special formatter for current column now.

Maybe I not so correct describe my task in issue, but I've task for when need only this decision.
I don't know how can describe this task another way

**Comment by cybermerlin** [comment]
PS
If u say @Select() to property then developer understand it like: "add for me please additional `SELECT * FROM`" - that's not correct for my task

**Comment by cybermerlin** [comment]
> I don't want to re-implement things when .selectAndMap will be introduced, 
This method is just for QueryBuilder if u write ur queryies in manual mode - is not my way.

File: src/driver/mysql/MysqlDriver.ts (hunk 3 above)
Comment (cybermerlin) [reply src/driver/mysql/MysqlDriver.ts]: mistakes, not repeated

**Comment by pleerock** [comment]
> If u say @select() to property then developer understand it like: "add for me please additional SELECT * FROM" - that's not correct for my task

**Comment by pleerock** [comment]
> If u say @select() to property then developer understand it like: "add for me please additional SELECT * FROM" - that's not correct for my task

that's not true. By this logic currently exist `addSelect` means it will add a new "SELECT * FROM".  `addSelect` just adds a new column/value for a selection. "SELECT * FROM" is called subquery.

> This method is just for QueryBuilder if u write ur queryies in manual mode - is not my way.

its not your way, but it fits ORM design. If you do something and want it to be part of ORM which used by many users be sure to fit everybody needs, not only your specific. As I said before I don't want to merge this and later re-implement everything again. It would be great if we have everything by design from beginning.

File: src/metadata/ColumnMetadata.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (pleerock) [reply src/metadata/ColumnMetadata.ts]: I don't see this supported by webstorm by default. Can you send link to official documentation for this feature? (for me it looks like a plugin or separately enabling thing). Also please send link for other ides (vscode, atom) for supporting this out of the box.

[the remark this answers is not in the document]
Comment (cybermerlin) [reply src/metadata/ColumnMetadata.ts]: [google](https://www.google.ru/search?newwindow=1&ei=RHW_WoSmIcb_sQG1_J-IAQ&q=code+folding+%2F%2Fregion+in+code&oq=code+folding+%2F%2Fregion+in+code&gs_l=psy-ab.3...349646.352000.0.353295.13.9.0.0.0.0.0.0..0.0....0...1c.1.64.psy-ab..13.0.0....0.VkUmMDnXGME)

[the remark this answers is not in the document]
Comment (cybermerlin) [reply src/metadata/ColumnMetadata.ts]: [idea jetbrains](https://www.jetbrains.com/help/idea/code-folding.html)

**Comment by Kononnable** [comment]
@cybermerlin Are you still interested in this PR? It's pretty old and if there are no plans for getting it up to date(or it would be easier to implement it again starting from current master) we should close it.

**Comment by cybermerlin** [comment]
@Kononnable, all changes in my fork, syncing with this repo would be not yet. In anythink mb closed. Because I want support webSQL, which it cutted in this repo, I'm not sure about apllyable to this repo my fork in future.

**Comment by imnotjames** [comment]
Sounds like you aren't interested in bringing this up to date.  That's ok!  If you do ever want to & feel comfortable reducing the scope of the changes so we can more easily review it please re-open a new PR.

Cheers!
